ChiModules/KEigenvalueSolver/IterativeMethods/power_iteration.cc

```diff
@@ -25,161 +28,128 @@ using namespace LinearBoltzmann;
  * Note that this routine currently only works when the problem
  * is defined by a single groupset.
 */
-void KEigenvalue::Solver::PowerIteration(LBSGroupset& groupset)
+void KEigenvalue::Solver::PowerIteration()
 {
   chi_log.Log(LOG_0)
-    << "\n\n";
-  chi_log.Log(LOG_0)
-    << "********** Solving k-eigenvalue problem with "
-    << "the Power Method.\n\n";
+      << "\n\n********** Solving k-eigenvalue problem with "
+      << "the Power Method.\n\n";
 
-  // ----- Setting up required sweep chunks
-  auto sweep_chunk = SetSweepChunk(groupset);
+  LBSGroupset& groupset = group_sets[0];
 
-  // ----- Set sweep scheduler
-  MainSweepScheduler SweepScheduler(SchedulingAlgorithm::DEPTH_OF_GRAPH,
-                                    groupset.angle_agg,
-                                    *sweep_chunk);
+  groupset.angle_agg.ZeroIncomingDelayedPsi();
 
-  // ----- Tool the sweep chunk
-  sweep_chunk->SetDestinationPhi(phi_new_local);
+  //======================================== Setup sweep chunk
+  auto sweep_chunk = SetSweepChunk(groupset);
+  MainSweepScheduler sweep_scheduler(SchedulingAlgorithm::DEPTH_OF_GRAPH,
+                                     groupset.angle_agg,
+                                     *sweep_chunk);
 
-  // ----- Set starting guess to a unit magnitude flux
-  phi_prev_local.assign(phi_prev_local.size(),1.0);
-  ScopedCopySTLvectors(groupset, phi_prev_local, phi_old_local);
+  //======================================== Tool the sweep chunk
+  sweep_scheduler.sweep_chunk.SetDestinationPhi(phi_new_local);
 
-  // ----- Start outer k iterations
-  double F_prev = 1.0;       //Production source prev
-  double k_eff_prev = k_eff;
-  int nit = 0;               //number of iterations
-  bool k_converged = false;
+  //======================================== Initial guess
+  phi_prev_local.assign(phi_prev_local.size(), 1.0);
+  ScopedCopySTLvectors(groupset, phi_prev_local, phi_old_local);
 
+  //======================================== Start power iterations
+  double F_prev = 1.0;
+  double k_eff_prev = 1.0;
+  int nit = 0;      //number of iterations
+  bool converged = false;
   while (nit < options.max_iterations)
   {
-    chi_log.Log(LOG_0VERBOSE_2)
-      << "\n********** Starting source iterations";
 
+    //============================== Clear source moments
+    q_moments_local.assign(q_moments_local.size(), 0.0);
+
+    //============================== Set the fission source
+    SetKSource(groupset, q_moments_local,
+               APPLY_AGS_FISSION_SOURCE |
+               APPLY_WGS_FISSION_SOURCE);
```

Comment:
Since the k fission source is simply preloaded into the q_moments_local vector, do we still need a third scalar flux vector? It seems like we can just use either `phi_old` or `phi_new` to first compute a fission integral, add a fission source to q_moments, converge a new iterate of 'phi_new', and finally compute a new fission integral to update `k`.

Also, should all the code relate to wgs scattering and ags scattering simply be removed from `SetKSource` since that function isn't meant to update scattering terms?

Reply:
I would revise this to be:

- `PowerIteration` iteration
    -  `SetKSource` -> `Q = F`
    - `Q_temp = Q`
    - `Richardson` or `GMRES` iterations
        -  `Q = Q_temp`
        - `SetSource` -> `Q = F + S + others`
        - Stuff

We store `Q_temp` before the inner iterations and then initialize `Q = Q_temp` in the inner iterations.

Reply:
@jthano This is the direction I want to take it, yes. I will likely make that change in the following PR which adds in groupsets to the k-eigenvalue solver. The primary purpose of this PR was to make the various `SetSource` routines write to a specified vector instead of clearing vectors within the routines.

Reply:
@Naktakala @jthano @zachhardy @dominiccaron @wdarylhawkins 
I like this discussion because I am learning something. Here's my thinking, which could be a concern. Those flags feel a little bit like a hack to me: "this is what happens to the source, provided that you have passed certain flags". I am afraid this can also be a source of errors or at least confusion. Is this the appropriate way of doing this?

Reply:
I agree with @ragusa. The flags infrastructure seems a bit error-prone to me. I have to admit though that I haven't had a chance to dig too deeply into the code, so I can't offer any constructive suggestions on what we might change or improve.  I'll make some time in the next day or so to take a look at things.

Reply:
Mmmmmm. Disappointing to see that it causes confusion only now because of the reflective BCs. Apart from the bug with BCs, the intent of using these flags was originally to help others reuse pieces of code, i.e., `SetSource`. And to that end, I believe it was kinda successful for these situations:
- The boolean options for GMRES simplified a lot, reminder GMRES calls `SetSource` in 2 places:
  - First to compute a source driven by material/boundary sources + Across-GroupSet (AGS) scattering
  - Then to compute a source driven by Within-GroupSet (WGS) scattering from a Krylov-vector
- Classic Richardson and GMRES became available to use by the KEigenvalue solver.
- Overloading `SetSource` became easier. You could add your own flags and both ClassicRichardson and GMRES didn't care.

I would like to hear from @jthano and @zachhardy on this topic since I believe they were the intended clients. Was the bit-wise flags a design mistake?

Reply:
I think this is something I'd like to discuss on Friday.

Reply:
I think the bit-wise flags are a good solution, they seem to be a popular and common tool for situations like this. I don't think the flags themselves are error prone, in fact I think they make the code more readable and less error prone.

It seems like there was just an odd code bug very specifically limited to reflecting boundaries that also happened to involve the flags.

Reply:
Personally, I like the bit-wise flags as well. Having many routines that do essentially the same thing as another with a tiny change (like dividing by k in the fission source) is suboptimal and the only other alternative I see to this is having several functions for the various components of the source. At the end of the day, those function names would incorporate the same information as the bit-wise flags. I say keep them.

Reply:
who said we should have many routines that are similar :-) ? As of now, I think I mostly need to find in the document what the list of flags are and make sure everybody is well educated of their existence and then use

Reply:
@ragusa Nobody said we had to but at the moment between LBS, the k-Eigenvalue Solver, and my transient solver (not in this repo) that is exactly what has happened. The bit-wise flags    have certainly taken us a step in the right direction away from duplicated code with minor changes.